// pulseaudio.js/dist/error.js
"use strict";
// Copyright (c) 2019-2021 Jan Janak <jan@janakj.org>
//
// Permission to use, copy, modify, and/or distribute this software for any
// purpose with or without fee is hereby granted, provided that the above
// copyright notice and this permission notice appear in all copies.
//
// THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
// WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
// MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
// ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
// WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
// ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
// OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.PulseError = exports.PA_ERR = void 0;
var PA_ERR;
(function (PA_ERR) {
    PA_ERR[PA_ERR["OK"] = 0] = "OK";
    PA_ERR[PA_ERR["ACCESS"] = 1] = "ACCESS";
    PA_ERR[PA_ERR["COMMAND"] = 2] = "COMMAND";
    PA_ERR[PA_ERR["INVALID"] = 3] = "INVALID";
    PA_ERR[PA_ERR["EXIST"] = 4] = "EXIST";
    PA_ERR[PA_ERR["NOENTITY"] = 5] = "NOENTITY";
    PA_ERR[PA_ERR["CONNECTIONREFUSED"] = 6] = "CONNECTIONREFUSED";
    PA_ERR[PA_ERR["PROTOCOL"] = 7] = "PROTOCOL";
    PA_ERR[PA_ERR["TIMEOUT"] = 8] = "TIMEOUT";
    PA_ERR[PA_ERR["AUTHKEY"] = 9] = "AUTHKEY";
    PA_ERR[PA_ERR["INTERNAL"] = 10] = "INTERNAL";
    PA_ERR[PA_ERR["CONNECTIONTERMINATED"] = 11] = "CONNECTIONTERMINATED";
    PA_ERR[PA_ERR["KILLED"] = 12] = "KILLED";
    PA_ERR[PA_ERR["INVALIDSERVER"] = 13] = "INVALIDSERVER";
    PA_ERR[PA_ERR["MODINITFAILED"] = 14] = "MODINITFAILED";
    PA_ERR[PA_ERR["BADSTATE"] = 15] = "BADSTATE";
    PA_ERR[PA_ERR["NODATA"] = 16] = "NODATA";
    PA_ERR[PA_ERR["VERSION"] = 17] = "VERSION";
    PA_ERR[PA_ERR["TOOLARGE"] = 18] = "TOOLARGE";
    PA_ERR[PA_ERR["NOTSUPPORTED"] = 19] = "NOTSUPPORTED";
    PA_ERR[PA_ERR["UNKNOWN"] = 20] = "UNKNOWN";
    PA_ERR[PA_ERR["NOEXTENSION"] = 21] = "NOEXTENSION";
    PA_ERR[PA_ERR["OBSOLETE"] = 22] = "OBSOLETE";
    PA_ERR[PA_ERR["NOTIMPLEMENTED"] = 23] = "NOTIMPLEMENTED";
    PA_ERR[PA_ERR["FORKED"] = 24] = "FORKED";
    PA_ERR[PA_ERR["IO"] = 25] = "IO";
    PA_ERR[PA_ERR["BUSY"] = 26] = "BUSY"; // Device or resource busy. \since 0.9.17
})(PA_ERR = exports.PA_ERR || (exports.PA_ERR = {}));
// Must match pulse/error.c
var paErrorToStr = (_a = {},
    _a[PA_ERR.OK] = "OK",
    _a[PA_ERR.ACCESS] = "Access denied",
    _a[PA_ERR.COMMAND] = "Unknown command",
    _a[PA_ERR.INVALID] = "Invalid argument",
    _a[PA_ERR.EXIST] = "Entity exists",
    _a[PA_ERR.NOENTITY] = "No such entity",
    _a[PA_ERR.CONNECTIONREFUSED] = "Connection refused",
    _a[PA_ERR.PROTOCOL] = "Protocol error",
    _a[PA_ERR.TIMEOUT] = "Timeout",
    _a[PA_ERR.AUTHKEY] = "No authentication key",
    _a[PA_ERR.INTERNAL] = "Internal error",
    _a[PA_ERR.CONNECTIONTERMINATED] = "Connection terminated",
    _a[PA_ERR.KILLED] = "Entity killed",
    _a[PA_ERR.INVALIDSERVER] = "Invalid server",
    _a[PA_ERR.MODINITFAILED] = "Module initialization failed",
    _a[PA_ERR.BADSTATE] = "Bad state",
    _a[PA_ERR.NODATA] = "No data",
    _a[PA_ERR.VERSION] = "Incompatible protocol version",
    _a[PA_ERR.TOOLARGE] = "Too large",
    _a[PA_ERR.NOTSUPPORTED] = "Not supported",
    _a[PA_ERR.UNKNOWN] = "Unknown error code",
    _a[PA_ERR.NOEXTENSION] = "No such extension",
    _a[PA_ERR.OBSOLETE] = "Obsolete functionality",
    _a[PA_ERR.NOTIMPLEMENTED] = "Missing implementation",
    _a[PA_ERR.FORKED] = "Client forked",
    _a[PA_ERR.IO] = "Input/Output error",
    _a[PA_ERR.BUSY] = "Device or resource busy",
    _a);
var PulseError = /** @class */ (function (_super) {
    __extends(PulseError, _super);
    function PulseError(packet) {
        var _this = this;
        var code = packet.getUInt32();
        _this = _super.call(this, paErrorToStr[code] || "Uknown Pulseaudio error code " + code) || this;
        _this.code = code;
        return _this;
    }
    return PulseError;
}(Error));
exports.PulseError = PulseError;
