// pulseaudio.js/dist/wav.js
"use strict";
// Copyright (c) 2019-2021 Jan Janak <jan@janakj.org>
//
// Permission to use, copy, modify, and/or distribute this software for any
// purpose with or without fee is hereby granted, provided that the above
// copyright notice and this permission notice appear in all copies.
//
// THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
// WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
// MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
// ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
// WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
// ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
// OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
Object.defineProperty(exports, "__esModule", { value: true });
exports.wavToChannelMap = void 0;
// Must match pulse/channelmap.{c,h}
var PA_CHANNEL;
(function (PA_CHANNEL) {
    PA_CHANNEL[PA_CHANNEL["POSITION_INVALID"] = -1] = "POSITION_INVALID";
    PA_CHANNEL[PA_CHANNEL["POSITION_MONO"] = 0] = "POSITION_MONO";
    PA_CHANNEL[PA_CHANNEL["POSITION_FRONT_LEFT"] = 1] = "POSITION_FRONT_LEFT";
    PA_CHANNEL[PA_CHANNEL["POSITION_FRONT_RIGHT"] = 2] = "POSITION_FRONT_RIGHT";
    PA_CHANNEL[PA_CHANNEL["POSITION_FRONT_CENTER"] = 3] = "POSITION_FRONT_CENTER";
    PA_CHANNEL[PA_CHANNEL["POSITION_REAR_CENTER"] = 4] = "POSITION_REAR_CENTER";
    PA_CHANNEL[PA_CHANNEL["POSITION_REAR_LEFT"] = 5] = "POSITION_REAR_LEFT";
    PA_CHANNEL[PA_CHANNEL["POSITION_REAR_RIGHT"] = 6] = "POSITION_REAR_RIGHT";
    PA_CHANNEL[PA_CHANNEL["POSITION_LFE"] = 7] = "POSITION_LFE";
    PA_CHANNEL[PA_CHANNEL["POSITION_FRONT_LEFT_OF_CENTER"] = 8] = "POSITION_FRONT_LEFT_OF_CENTER";
    PA_CHANNEL[PA_CHANNEL["POSITION_FRONT_RIGHT_OF_CENTER"] = 9] = "POSITION_FRONT_RIGHT_OF_CENTER";
    PA_CHANNEL[PA_CHANNEL["POSITION_SIDE_LEFT"] = 10] = "POSITION_SIDE_LEFT";
    PA_CHANNEL[PA_CHANNEL["POSITION_SIDE_RIGHT"] = 11] = "POSITION_SIDE_RIGHT";
    PA_CHANNEL[PA_CHANNEL["POSITION_AUX0"] = 12] = "POSITION_AUX0";
    PA_CHANNEL[PA_CHANNEL["POSITION_AUX1"] = 13] = "POSITION_AUX1";
    PA_CHANNEL[PA_CHANNEL["POSITION_AUX2"] = 14] = "POSITION_AUX2";
    PA_CHANNEL[PA_CHANNEL["POSITION_AUX3"] = 15] = "POSITION_AUX3";
    PA_CHANNEL[PA_CHANNEL["POSITION_AUX4"] = 16] = "POSITION_AUX4";
    PA_CHANNEL[PA_CHANNEL["POSITION_AUX5"] = 17] = "POSITION_AUX5";
    PA_CHANNEL[PA_CHANNEL["POSITION_AUX6"] = 18] = "POSITION_AUX6";
    PA_CHANNEL[PA_CHANNEL["POSITION_AUX7"] = 19] = "POSITION_AUX7";
    PA_CHANNEL[PA_CHANNEL["POSITION_AUX8"] = 20] = "POSITION_AUX8";
    PA_CHANNEL[PA_CHANNEL["POSITION_AUX9"] = 21] = "POSITION_AUX9";
    PA_CHANNEL[PA_CHANNEL["POSITION_AUX10"] = 22] = "POSITION_AUX10";
    PA_CHANNEL[PA_CHANNEL["POSITION_AUX11"] = 23] = "POSITION_AUX11";
    PA_CHANNEL[PA_CHANNEL["POSITION_AUX12"] = 24] = "POSITION_AUX12";
    PA_CHANNEL[PA_CHANNEL["POSITION_AUX13"] = 25] = "POSITION_AUX13";
    PA_CHANNEL[PA_CHANNEL["POSITION_AUX14"] = 26] = "POSITION_AUX14";
    PA_CHANNEL[PA_CHANNEL["POSITION_AUX15"] = 27] = "POSITION_AUX15";
    PA_CHANNEL[PA_CHANNEL["POSITION_AUX16"] = 28] = "POSITION_AUX16";
    PA_CHANNEL[PA_CHANNEL["POSITION_AUX17"] = 29] = "POSITION_AUX17";
    PA_CHANNEL[PA_CHANNEL["POSITION_AUX18"] = 30] = "POSITION_AUX18";
    PA_CHANNEL[PA_CHANNEL["POSITION_AUX19"] = 31] = "POSITION_AUX19";
    PA_CHANNEL[PA_CHANNEL["POSITION_AUX20"] = 32] = "POSITION_AUX20";
    PA_CHANNEL[PA_CHANNEL["POSITION_AUX21"] = 33] = "POSITION_AUX21";
    PA_CHANNEL[PA_CHANNEL["POSITION_AUX22"] = 34] = "POSITION_AUX22";
    PA_CHANNEL[PA_CHANNEL["POSITION_AUX23"] = 35] = "POSITION_AUX23";
    PA_CHANNEL[PA_CHANNEL["POSITION_AUX24"] = 36] = "POSITION_AUX24";
    PA_CHANNEL[PA_CHANNEL["POSITION_AUX25"] = 37] = "POSITION_AUX25";
    PA_CHANNEL[PA_CHANNEL["POSITION_AUX26"] = 38] = "POSITION_AUX26";
    PA_CHANNEL[PA_CHANNEL["POSITION_AUX27"] = 39] = "POSITION_AUX27";
    PA_CHANNEL[PA_CHANNEL["POSITION_AUX28"] = 40] = "POSITION_AUX28";
    PA_CHANNEL[PA_CHANNEL["POSITION_AUX29"] = 41] = "POSITION_AUX29";
    PA_CHANNEL[PA_CHANNEL["POSITION_AUX30"] = 42] = "POSITION_AUX30";
    PA_CHANNEL[PA_CHANNEL["POSITION_AUX31"] = 43] = "POSITION_AUX31";
    PA_CHANNEL[PA_CHANNEL["POSITION_TOP_CENTER"] = 44] = "POSITION_TOP_CENTER";
    PA_CHANNEL[PA_CHANNEL["POSITION_TOP_FRONT_LEFT"] = 45] = "POSITION_TOP_FRONT_LEFT";
    PA_CHANNEL[PA_CHANNEL["POSITION_TOP_FRONT_RIGHT"] = 46] = "POSITION_TOP_FRONT_RIGHT";
    PA_CHANNEL[PA_CHANNEL["POSITION_TOP_FRONT_CENTER"] = 47] = "POSITION_TOP_FRONT_CENTER";
    PA_CHANNEL[PA_CHANNEL["POSITION_TOP_REAR_LEFT"] = 48] = "POSITION_TOP_REAR_LEFT";
    PA_CHANNEL[PA_CHANNEL["POSITION_TOP_REAR_RIGHT"] = 49] = "POSITION_TOP_REAR_RIGHT";
    PA_CHANNEL[PA_CHANNEL["POSITION_TOP_REAR_CENTER"] = 50] = "POSITION_TOP_REAR_CENTER";
    PA_CHANNEL[PA_CHANNEL["POSITION_MAX"] = 51] = "POSITION_MAX";
})(PA_CHANNEL || (PA_CHANNEL = {}));
function wavToChannelMap(channels) {
    /* Following http://www.microsoft.com/whdc/device/audio/multichaud.mspx#EKLAC */
    var map = Array(channels);
    switch (channels) {
        case 1:
            map[0] = PA_CHANNEL.POSITION_MONO;
            return map;
        case 18:
            map[15] = PA_CHANNEL.POSITION_TOP_REAR_LEFT;
            map[16] = PA_CHANNEL.POSITION_TOP_REAR_CENTER;
            map[17] = PA_CHANNEL.POSITION_TOP_REAR_RIGHT;
        /* Fall through */
        case 15:
            map[12] = PA_CHANNEL.POSITION_TOP_FRONT_LEFT;
            map[13] = PA_CHANNEL.POSITION_TOP_FRONT_CENTER;
            map[14] = PA_CHANNEL.POSITION_TOP_FRONT_RIGHT;
        /* Fall through */
        case 12:
            map[11] = PA_CHANNEL.POSITION_TOP_CENTER;
        /* Fall through */
        case 11:
            map[9] = PA_CHANNEL.POSITION_SIDE_LEFT;
            map[10] = PA_CHANNEL.POSITION_SIDE_RIGHT;
        /* Fall through */
        case 9:
            map[8] = PA_CHANNEL.POSITION_REAR_CENTER;
        /* Fall through */
        case 8:
            map[6] = PA_CHANNEL.POSITION_FRONT_LEFT_OF_CENTER;
            map[7] = PA_CHANNEL.POSITION_FRONT_RIGHT_OF_CENTER;
        /* Fall through */
        case 6:
            map[4] = PA_CHANNEL.POSITION_REAR_LEFT;
            map[5] = PA_CHANNEL.POSITION_REAR_RIGHT;
        /* Fall through */
        case 4:
            map[3] = PA_CHANNEL.POSITION_LFE;
        /* Fall through */
        case 3:
            map[2] = PA_CHANNEL.POSITION_FRONT_CENTER;
        /* Fall through */
        case 2:
            map[0] = PA_CHANNEL.POSITION_FRONT_LEFT;
            map[1] = PA_CHANNEL.POSITION_FRONT_RIGHT;
            return map;
        default:
            throw new Error("Unsupported number of channels " + channels);
    }
}
exports.wavToChannelMap = wavToChannelMap;
