// @tmigone/pulseaudio/build/tag.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PAU8 = exports.PAFormat = exports.PAVolume = exports.PAUsec = exports.PAChannelVolume = exports.PAChannelMap = exports.PASampleSpec = exports.PAPropList = exports.PAProp = exports.PAString = exports.PAU32 = exports.PABoolean = exports.PAArbitrary = exports.PATagType = exports.PATag = void 0;
const common_1 = require("./tags/common");
Object.defineProperty(exports, "PATag", { enumerable: true, get: function () { return common_1.PATag; } });
Object.defineProperty(exports, "PATagType", { enumerable: true, get: function () { return common_1.PATagType; } });
const u32_1 = require("./tags/u32");
exports.PAU32 = u32_1.default;
const arbitrary_1 = require("./tags/arbitrary");
exports.PAArbitrary = arbitrary_1.default;
const boolean_1 = require("./tags/boolean");
exports.PABoolean = boolean_1.default;
const string_1 = require("./tags/string");
exports.PAString = string_1.default;
const prop_1 = require("./tags/prop");
exports.PAProp = prop_1.default;
const propList_1 = require("./tags/propList");
exports.PAPropList = propList_1.default;
const sampleSpec_1 = require("./tags/sampleSpec");
exports.PASampleSpec = sampleSpec_1.default;
const channelMap_1 = require("./tags/channelMap");
exports.PAChannelMap = channelMap_1.default;
const channelVolume_1 = require("./tags/channelVolume");
exports.PAChannelVolume = channelVolume_1.default;
const usec_1 = require("./tags/usec");
exports.PAUsec = usec_1.default;
const volume_1 = require("./tags/volume");
exports.PAVolume = volume_1.default;
const format_1 = require("./tags/format");
exports.PAFormat = format_1.default;
const u8_1 = require("./tags/u8");
exports.PAU8 = u8_1.default;
